packages/indexer/src/plugins/kv.ts

```diff
@@ -1,16 +1,112 @@
+import assert from "node:assert";
+import type { Cursor, DataFinality } from "@apibara/protocol";
+import { type Database, type ISqlite, open } from "sqlite";
 import { useIndexerContext } from "../context";
-import { defineIndexerPlugin } from "../plugins";
+import { deserialize, serialize } from "../vcr";
+import { defineIndexerPlugin } from "./config";
 
-/** This plugin is a placeholder for a future key-value store plugin. */
-export function kv<TFilter, TBlock, TRet>() {
+type SqliteArgs = ISqlite.Config;
+
+export function kv<TFilter, TBlock, TRet>(args: SqliteArgs) {
   return defineIndexerPlugin<TFilter, TBlock, TRet>((indexer) => {
+    let db: Database;
+
     indexer.hooks.hook("run:before", async () => {
+      db = await open(args);
+      await db.exec(`
+        CREATE TABLE IF NOT EXISTS kvs (
+          from_block INTEGER NOT NULL,
+          to_block INTEGER,
+          k TEXT NOT NULL,
+          v BLOB NOT NULL,
+          PRIMARY KEY (from_block, k)
+        );
+      `);
+    });
+
+    indexer.hooks.hook("handler:before", async ({ finality, endCursor }) => {
+      const ctx = useIndexerContext();
+
+      assert(endCursor, new Error("endCursor cannot be undefined"));
+
+      ctx.kv = new KVStore(db, finality, endCursor);
+
+      await db.exec("BEGIN TRANSACTION");
```

Comment:
My advice is to put this into a function on the kv so that it can be tested. You should test both the successfull flow (begin tx -> commit) and the exception flow (begin tx -> rollback)

```suggestion
      ctx.kv = new KVStore(db, finality, endCursor);

      await ctx.kv.beginTransaction();
```